### PR TITLE
Add Initial TE-M3200 pressure sensor support

### PR DESCRIPTION
**Pull request by bakerkj**
# What does this implement/fix?

Adds support for the TE-M3200 i2c pressure sensor.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** none

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3915

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
```yaml
# Example config.yaml
i2c:
  id: i2c_bus
  scan: true
  sda: 21
  scl: 22
  frequency: 200kHz

sensor:
  - platform: tem3200
    i2c_id: i2c_bus
    temperature:
      name: Temperature
    raw_pressure:
      id: raw_pressure
      internal: true

  - platform: copy
    source_id: raw_pressure
    name: Pressure
    state_class: measurement
    device_class: pressure
    unit_of_measurement: psi
    accuracy_decimals: 2
    filters:
      # 100 psi sensor
      - calibrate_linear:
          - 1000 -> 0.0
          - 8000 -> 50.0
          - 15000 -> 100.0
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/tem3200/tem3200.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  const float temperature_min = -50;
+  const float temperature_span = temperature_max - temperature_min;
+
+  float temperature = (raw_temperature * temperature_span / temperature_bits_span) + temperature_min;
+
+  return temperature;
+}
+
+void TEM3200Component::update() {
+  this->set_timeout(50, [this]() {
```
Comment (swoboda1337) [inline esphome/components/tem3200/tem3200.cpp]: Why do you need this timeout when nothing is done before it? What are you waiting for?

Comment (bakerkj) [reply esphome/components/tem3200/tem3200.cpp]: @swoboda1337, I duplicated this from another component I was modeling some of my work on. Given your comment, I believe I misunderstood its purpose. There is nothing before this to wait on. Given your comment, and reading some more, I believe it is not needed.

Comment (bakerkj) [reply esphome/components/tem3200/tem3200.cpp]: I have updated the code to remove this timeout. Thanks for your help!

File: esphome/components/tem3200/sensor.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    CONF_ID,
+    CONF_TEMPERATURE,
+    DEVICE_CLASS_TEMPERATURE,
+    STATE_CLASS_MEASUREMENT,
+    UNIT_CELSIUS,
+)
+
+CODEOWNERS = ["@bakerkj"]
+DEPENDENCIES = ["i2c"]
+CONFIG_MULTI = True
```
Comment (jesserockz) [inline esphome/components/tem3200/sensor.py]: Not required
```suggestion
```

File: esphome/components/tem3200/sensor.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+
+    if CONF_TEMPERATURE in config:
+        sens = await sensor.new_sensor(config[CONF_TEMPERATURE])
```
Comment (jesserockz) [inline esphome/components/tem3200/sensor.py]:
```suggestion
    if temperature_config := config.get(CONF_TEMPERATURE):
        sens = await sensor.new_sensor(temperature_config)
```

File: esphome/components/tem3200/sensor.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        cg.add(var.set_temperature_sensor(sens))
+
+    if CONF_RAW_PRESSURE in config:
+        sens = await sensor.new_sensor(config[CONF_RAW_PRESSURE])
```
Comment (jesserockz) [inline esphome/components/tem3200/sensor.py]:
```suggestion
    if raw_pressure_config := config.get(CONF_RAW_PRESSURE):
        sens = await sensor.new_sensor(raw_pressure_config)
```

File: esphome/components/tem3200/tem3200.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+#include "esphome/components/sensor/sensor.h"
+#include "esphome/components/i2c/i2c.h"
+
+namespace esphome {
+namespace tem3200 {
+
+/// This class implements support for the tem3200 pressure and temperature i2c sensors.
+class TEM3200Component : public PollingComponent, public i2c::I2CDevice {
+ public:
+  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { temperature_sensor_ = temperature_sensor; }
```
Comment (jesserockz) [inline esphome/components/tem3200/tem3200.h]: https://esphome.io/guides/contributing#codebase-standards
> All uses of class members and member functions should be prefixed with this-> to distinguish them from global functions in code review.

```suggestion
  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { this->temperature_sensor_ = temperature_sensor; }
```

File: esphome/components/tem3200/tem3200.cpp [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  // extract top 6 bits of first byte and all bits of second byte for pressure
+  raw_pressure = (((response[0] & 0x3f)) << 8 | response[1]);
+
+  // extract all bytes of 3rd byte and top 3 bits of fourth byte for temperature
+  raw_temperature = ((response[2] << 3) | (response[3] & 0xe0) >> 5);
+
+  return i2c::ERROR_OK;
+}
+
+float TEM3200Component::convert_temperature_(uint16_t raw_temperature) {
```
Comment (jesserockz) [inline esphome/components/tem3200/tem3200.cpp]: This function can probably just be an inline function outside of the class itself.
```suggestion
inline float convert_temperature(uint16_t raw_temperature) {
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/tem3200/tem3200.cpp (hunk 6 above)
Comment (bakerkj) [reply esphome/components/tem3200/tem3200.cpp]: fixed.

**Review by jesserockz — APPROVED** [review]

File: esphome/components/tem3200/tem3200.cpp [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+      return;
+    case STALE:
+      ESP_LOGE(TAG, "Warning: STALE data. Data has not been updated since last fetch");
+      this->status_set_warning();
+      return;
+  }
+
+  float temperature = convert_temperature(raw_temperature);
+
+  ESP_LOGD(TAG, "Got pressure=%draw temperature=%.1f°C", raw_pressure, temperature);
```
Comment (jesserockz) [inline esphome/components/tem3200/tem3200.cpp]:
```suggestion
  ESP_LOGD(TAG, "Got raw pressure=%d, temperature=%.1f°C", raw_pressure, temperature);
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Comment by bakerkj** [comment]
@jesserockz thank you for your help and approvals. Is there anything else I need to do to merge this in?

**Pull request merged**